### PR TITLE
Commands for adding/removing infinispan caches

### PR DESCRIPTION
**Pull request by jmartisk**
This also adds support for Double in Values and ModelNodeResult, because it is needed for some properties of distributed caches.

Offline version of commands... Ugh, maybe later, aligator.

### CODE REVIEW

File: commands/src/main/java/org/wildfly/extras/creaper/commands/infinispan/cache/AbstractAddCache.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+import org.wildfly.extras.creaper.core.online.OnlineCommandContext;
+import org.wildfly.extras.creaper.core.online.operations.Address;
+import org.wildfly.extras.creaper.core.online.operations.Operations;
+import org.wildfly.extras.creaper.core.online.operations.Values;
+
+
+/**
+ * @author Jan Martiska
+ */
+public abstract class AbstractAddCache implements OnlineCommand {
```
Comment (Ladicek) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/infinispan/cache/AbstractAddCache.java]: We don't do `@author`s in Creaper. It's visible from Git history.

Also, I don't think this class should be `public`.

File: commands/src/main/java/org/wildfly/extras/creaper/commands/infinispan/cache/AbstractAddCache.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        this.jndiName = builder.jndiName;
+        this.module = builder.module;
+        this.start = builder.start;
+        this.statisticsEnabled = builder.statisticsEnabled;
+        this.address = Address.subsystem("infinispan")
+                .and("cache-container", cacheContainer)
+                .and(cacheType.getType(), name);
+    }
+
+    public static class Builder<THIS extends Builder> {
```
Comment (Ladicek) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/infinispan/cache/AbstractAddCache.java]: This should probably be `abstract`.

File: commands/src/main/java/org/wildfly/extras/creaper/commands/infinispan/cache/AbstractAddCache.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+        String name;
+        String cacheContainer;
+        String jndiName;
+        String module;
+        String start;
+        Boolean statisticsEnabled;
```
Comment (Ladicek) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/infinispan/cache/AbstractAddCache.java]: If these are supposed to be visible by subclasses, I'd make them `protected` (though it's not needed).

File: commands/src/main/java/org/wildfly/extras/creaper/commands/infinispan/cache/AbstractAddCache.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+        public Builder(String name) {
```
Comment (Ladicek) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/infinispan/cache/AbstractAddCache.java]: Should probably be `protected`.

File: commands/src/main/java/org/wildfly/extras/creaper/commands/infinispan/cache/AddDistributedCache.java [hunk 5]
```diff
@@ -0,0 +1,115 @@
+package org.wildfly.extras.creaper.commands.infinispan.cache;
+
+import org.wildfly.extras.creaper.core.online.operations.Values;
+
+/**
+ * @author Jan Martiska
+ */
+public class AddDistributedCache extends AbstractAddCache {
```
Comment (Ladicek) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/infinispan/cache/AddDistributedCache.java]: `final`

File: commands/src/main/java/org/wildfly/extras/creaper/commands/infinispan/cache/AddDistributedCache.java [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    private final CacheMode mode;
+    private final Long queueFlushInterval;
+    private final Long remoteTimeout;
+    private final Double capacityFactor;
+    private final ConsistentHashStrategy consistentHashStrategy;
+    private final Long l1lifespan;
+    private final Integer owners;
+    private final Integer segments;
+
+    public AddDistributedCache(Builder builder) {
```
Comment (Ladicek) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/infinispan/cache/AddDistributedCache.java]: `private`, should only be called from the builder.

File: commands/src/main/java/org/wildfly/extras/creaper/commands/infinispan/cache/AddDistributedCache.java [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+                .andOptional("remote-timeout", remoteTimeout)
+                .andOptional("capacity-factor", capacityFactor)
+                .andOptional("consistent-hash-strategy",
+                        consistentHashStrategy != null ? consistentHashStrategy.getType() : null)
+                .andOptional("l1-lifespan", l1lifespan)
+                .andOptional("owners", owners)
+                .andOptional("segments", segments);
+    }
+
+    public static class Builder extends AbstractAddCache.Builder<Builder> {
```
Comment (Ladicek) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/infinispan/cache/AddDistributedCache.java]: `final`

File: commands/src/main/java/org/wildfly/extras/creaper/commands/infinispan/cache/RemoveCache.java [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+import org.wildfly.extras.creaper.core.online.OnlineCommand;
+import org.wildfly.extras.creaper.core.online.OnlineCommandContext;
+import org.wildfly.extras.creaper.core.online.operations.Address;
+import org.wildfly.extras.creaper.core.online.operations.Operations;
+
+/**
+ * @author Jan Martiska
+ */
+public class RemoveCache implements OnlineCommand {
```
Comment (Ladicek) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/infinispan/cache/RemoveCache.java]: `final`

File: core/src/main/java/org/wildfly/extras/creaper/core/online/operations/Values.java [hunk 9]
```diff
@@ -157,6 +161,12 @@ public Values and(String name, String value) {
         return new Values(newList);
     }
 
+    public Values and(String name, Double value) {
```
Comment (Ladicek) [inline core/src/main/java/org/wildfly/extras/creaper/core/online/operations/Values.java]: This should use primitive `double`, similarly to other primitive types.

File: core/src/main/java/org/wildfly/extras/creaper/core/online/operations/Values.java [hunk 10]
```diff
@@ -243,9 +258,21 @@ public Values andList(String name, ModelNode... value) {
         return new Values(newList);
     }
 
+    public Values andList(String name, Double... value) {
```
Comment (Ladicek) [inline core/src/main/java/org/wildfly/extras/creaper/core/online/operations/Values.java]: This should use primitive `double`, similarly to other primitive types.

File: commands/src/main/java/org/wildfly/extras/creaper/commands/infinispan/cache/AddDistributedCache.java [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+    @Override
+    final Values addValuesSpecificForCacheType(Values generalCacheValues) {
+        return generalCacheValues
+                .andOptional("async-marshalling", asyncMarshalling)
+                .andOptional("mode", mode != null ? mode.getMode() : null)
+                .andOptional("queue-flush-interval", queueFlushInterval)
+                .andOptional("remote-timeout", remoteTimeout)
+                .andOptional("capacity-factor", capacityFactor)
+                .andOptional("consistent-hash-strategy",
+                        consistentHashStrategy != null ? consistentHashStrategy.getType() : null)
```
Comment (Ladicek) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/infinispan/cache/AddDistributedCache.java]: The test says that `capacity-factor` and `consistent-hash-strategy` are WFLY 9+ only -- if so, we should throw an exception if they are set and current server is < WFLY 9. Or silently ignore them, but I think that we typically do the exception thing.

**Comment by Ladicek** [comment]
All comments for `AddDistributedCache` also apply to other `Add*Cache` commands.

Generally, nice work, the comments are all minor, I don't think this will require any additional substantial work, just stylistics.

**Comment by Ladicek** [comment]
BTW, offline variants are _not_ required.

File: commands/src/main/java/org/wildfly/extras/creaper/commands/infinispan/cache/AddDistributedCache.java (hunk 11 above)
Comment (jmartisk) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/infinispan/cache/AddDistributedCache.java]: Personally, throwing an exception doesn't seem quite right to me. The test would have to contain `if`s to decide whether to add the params or not. And wildfly doesn't complain either, when you pass a parameter which doesn't do anything. Wouldn't a WARN log do a better job?

Comment (Ladicek) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/infinispan/cache/AddDistributedCache.java]: I'm fine with skipping those parameters, if the builder methods have a javadoc stating something like "{@code capacity-factor} is only supported on WildFly 9 and above, this setting will be ignored on previous versions".

Comment (jmartisk) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/infinispan/cache/AddDistributedCache.java]: Ok, that sounds reasonable.

**Comment by jmartisk** [comment]
I did the changes as you requested, I hope I didn't forget anything.
The comments about unsupported params currently only apply to distributed caches, I didn't find any other parameters which aren't supported (tried all the way back to EAP 6.4).

**Comment by Ladicek** [comment]
I've made a couple of stylistic changes myself and pushed. Thanks!

**Pull request closed without merging**